patroni/dcs/__init__.py

```diff
@@ -85,38 +87,79 @@ def dcs_modules() -> List[str]:
         return [module_prefix + name for _, name, is_pkg in pkgutil.iter_modules([dcs_dirname]) if not is_pkg]
 
 
-def get_dcs(config: Union['Config', Dict[str, Any]]) -> 'AbstractDCS':
-    modules = dcs_modules()
+def iter_dcs_classes(
+        config: Optional[Union['Config', Dict[str, Any]]] = None
+) -> Iterator[Tuple[str, Optional[Type['AbstractDCS']]]]:
+    """Attempt to import DCS modules that are present in the given configuration.
+
+    .. note::
+            If a module successfully imports we can assume that all its requirements are installed.
+
+    :param config: configuration information with possible DCS names as keys. If given, only attempt to import DCS
+                   modules defined in the configuration. Else, if ``None``, attempt to import any supported DCS module.
 
-    for module_name in modules:
-        name = module_name.split('.')[-1]
-        if name in config:  # we will try to import only modules which have configuration section in the config file
+    :yields: a tuple containing the module ``name`` and the imported DCS class object.
+    """
+    for mod_name in dcs_modules():
+        name = mod_name.rpartition('.')[2]
+        if config is None or name in config:
             try:
-                module = importlib.import_module(module_name)
-                for key, item in module.__dict__.items():  # iterate through the module content
-                    # try to find implementation of AbstractDCS interface, class name must match with module_name
-                    if key.lower() == name and inspect.isclass(item) and issubclass(item, AbstractDCS):
-                        # propagate some parameters
-                        config[name].update({p: config[p] for p in ('namespace', 'name', 'scope', 'loop_wait',
-                                             'patronictl', 'ttl', 'retry_timeout') if p in config})
-                        # From citus section we only need "group" parameter, but will propagate everything just in case.
-                        if isinstance(config.get('citus'), dict):
-                            config[name].update(config['citus'])
-                        return item(config[name])
+                module = importlib.import_module(mod_name)
+                yield name, find_dcs_class_in_module(module)
             except ImportError:
-                logger.debug('Failed to import %s', module_name)
+                logger.info('Failed to import %s', mod_name)
```

Comment:
Still a problem.
Now it will show unwanted messages when for example we have `kubernetes` in the configuration and no `python-etcd`, `python-consul`, `kazoo`, or `pysyncobj` modules installed.

Reply:
I'll add a condition for the log level based on whether the `name` is in `config`. INFO for configured, DEBUG otherwise.

Reply:
This is what the original code did:
```python
    available_implementations: List[str] = []
    for module_name in modules:
        name = module_name.split('.')[-1]
        try:
            module = importlib.import_module(module_name)
            available_implementations.extend(name for key, item in module.__dict__.items() if key.lower() == name
                                             and inspect.isclass(item) and issubclass(item, AbstractDCS))
        except ImportError:
            logger.info('Failed to import %s', module_name)
```

It is importing modules and trying to find a class derived from `AbstractDCS` and with the `__class__.name.lower()` matching the module_name`.
That is, the empty file `patroni/dcs/foo.py` will be skipped.

Reply:
OK, I've pushed a couple of commits which I hope address this. `foo` is skipped now.

I reverted to passing back `name, dcs_class` instead of using `__class__.__name__`. There was a test which conflicted with this which I have added a suggested workaround. It might need a `patch()` instead? I also thought there is less clutter to using the dunder attrs.